src/uct/api/uct.h

```diff
@@ -2063,25 +2064,136 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
+     * filled with the actual size that was allocated. Must be >0.
+     */
+    size_t             *length_p;
+
+    /**
+     * Type of memory to be allocated.
+     */
+    ucs_memory_type_t  mem_type;
+
+    struct {
+        /**
+         * Array of memory allocation methods to attempt. Each of the provided
+         * allocation methods will be used, in turn, to perform the allocation,
+         * until one succeeds. Whenever the MD method is encountered, each of
+         * the provided MDs will be used, in turn, to allocate the memory, until
+         * one succeeds, or they are exhausted. In this case the next allocation
+         * method from the initial list will be attempted.
+         */
+        const uct_alloc_method_t *methods;
+
+        /**
+         * Length of 'methods' array
+         */
+        unsigned                 count;
+    } methods;
+
+    struct {
+        /**
+         * Array of memory domains to attempt to allocate
+         * the memory with, for MD allocation method.
+         */
+        const uct_md_h *mds;
+
+        /**
+         *  Length of 'mds' array. May be empty, in such case
+         *  'mds' may be NULL, and MD allocation method will
+         *  be skipped.
+         */
+        unsigned       count;
+    } mds;
+
+    /**
+     * Name of the allocated region, used to track memory
+     * usage for debugging and profiling.
+     */
+    const char         *name;
+} uct_mem_alloc_params_t;
+
+
 /**
  * @ingroup UCT_MD
  * @brief Allocate memory for zero-copy sends and remote access.
  *
  * Allocate memory on the memory domain. In order to use this function, MD
  * must support @ref UCT_MD_FLAG_ALLOC flag.
  *
- * @param [in]     md          Memory domain to allocate memory on.
- * @param [in,out] length_p    Points to the size of memory to allocate. Upon successful
- *                             return, filled with the actual size that was allocated,
- *                             which may be larger than the one requested. Must be >0.
- * @param [in,out] address_p   The address
- * @param [in]     flags       Memory allocation flags, see @ref uct_md_mem_flags.
- * @param [in]     name        Name of the allocated region, used to track memory
- *                             usage for debugging and profiling.
+ * @param [in]     param       Memory allocation characteristics, see
+ *                             @ref uct_mem_alloc_params_t.
  * @param [out]    memh_p      Filled with handle for allocated region.
+ *
+ * Note: param->methods.count must be 1, param->methods.method must be
+ * UCT_ALLOC_METHOD_MD, param->mds.count must be 1, and param->mds.md
+ * must point to memory domain on which to allocate
  */
-ucs_status_t uct_md_mem_alloc(uct_md_h md, size_t *length_p, void **address_p,
-                              unsigned flags, const char *name, uct_mem_h *memh_p);
+ucs_status_t uct_md_mem_alloc(const uct_mem_alloc_params_t *param,
+                              uct_mem_h *memh_p);
```

Comment:
we should not reuse same params struct for both uct_md_mem_alloc() and uct_mem_alloc(). We can remove uct_md_mem_alloc() completely from UCT API and keep only the generic one, uct_mem_alloc(). Then, we can add regular mem_type parameter to uct_md_mem_alloc() as an internal function.